### PR TITLE
SIMSBIOHUB-587 API: Bulk Captures

### PR DESCRIPTION
**Pull request by MacDeluca**
## Links to Jira Tickets

- [SIMSBIOHUB-587](https://apps.nrs.gov.bc.ca/int/jira/secure/RapidBoard.jspa?rapidView=11325&projectKey=SIMSBIOHUB&view=detail&selectedIssue=SIMSBIOHUB-587#)

## Description of Changes (only backend)

- Bulk import Critterbase captures via CSV
- importCSV strategy function
- updated columnValidator spec
- import services now implement shared interface (to be injected into importCSV)
- column validator now handles optional properties

## Testing Notes

- Should be able to import capture CSV file with postman

### CODE REVIEW

**Comment by NickPhura** [comment]
Regarding the `csvFileSchema` allowed mime types enum `[text/csv]`:

I'm encountering an interesting issue, where the mime type, despite the extension being `.csv`, is parsed as `application/vnd.ms-excel`.

This blog post actually mentions the same issue: https://christianwood.net/posts/csv-file-upload-validation/

Their conclusion that they just have to add more mime types to the allowed list.

The issue doesn't seem to happen through postman.

(Everything else downstream works, I just had to add that extra mime type to the openapi schema)

File: api/src/paths/project/{projectId}/survey/{surveyId}/critters/{critterId}/captures/import.ts [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+      in: 'path',
+      name: 'surveyId',
+      required: true
+    },
+    {
+      in: 'path',
+      description: 'SIMS survey critter id.',
+      name: 'critterId',
+      required: true
+    }
```
Comment (NickPhura) [inline api/src/paths/project/{projectId}/survey/{surveyId}/critters/{critterId}/captures/import.ts]: nit: specify type (integer, etc)

File: api/src/services/import-services/csv-import-strategy.interface.ts [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+   * @see '../../utils/xlsx-utils/column-cell-utils.ts'
+   */
+  columnValidator: IXLSXCSVValidator;
+
+  /**
+   * Validate the pre-parsed rows - return either custom Validation or Zod SafeParse.
+   *
+   * @param {Row[]} rows - Raw unparsed CSV rows
+   * @param {WorkSheet} [worksheet] - Xlsx worksheet - useful for calculating non-standard columns
+   * @returns {*} Validation
```
Comment (NickPhura) [inline api/src/services/import-services/csv-import-strategy.interface.ts]: nit: return looks like it needs updating

File: api/src/services/import-services/csv-import-strategy.interface.ts [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+ *
+ */
+export type Row = Record<string, any>;
+
+/**
+ * Implementation for CSV Import Services.
+ *
+ * @description All CSV import services should implement this interface to be used with `CSVImportStrategy`
+ */
+export interface CSVImportService<ValidatedRow = Record<string, unknown>, InsertReturn = unknown> {
```
Comment (NickPhura) [inline api/src/services/import-services/csv-import-strategy.interface.ts:20]: nit: could add the @template jsdoc for the 2 generics

File: api/src/paths/project/{projectId}/survey/{surveyId}/critters/{critterId}/captures/import.ts (hunk 1 above)
Comment (MacDeluca) [reply api/src/paths/project/{projectId}/survey/{surveyId}/critters/{critterId}/captures/import.ts]: fixed

File: api/src/services/import-services/csv-import-strategy.interface.ts (hunk 2 above)
Comment (MacDeluca) [reply api/src/services/import-services/csv-import-strategy.interface.ts]: fixed

File: api/src/services/import-services/csv-import-strategy.interface.ts (hunk 3 above)
Comment (MacDeluca) [reply api/src/services/import-services/csv-import-strategy.interface.ts:20]: fixed

**Comment by MacDeluca** [comment]
> Regarding the `csvFileSchema` allowed mime types enum `[text/csv]`:
> 
> I'm encountering an interesting issue, where the mime type, despite the extension being `.csv`, is parsed as `application/vnd.ms-excel`.
> 
> This blog post actually mentions the same issue: https://christianwood.net/posts/csv-file-upload-validation/
> 
> Their conclusion that they just have to add more mime types to the allowed list.
> 
> The issue doesn't seem to happen through postman.
> 
> (Everything else downstream works, I just had to add that extra mime type to the openapi schema)

I have updated the file schema to include all of these. I can remove some if we think they are not needed.

File: api/src/paths/project/{projectId}/survey/{surveyId}/critters/import.test.ts [hunk 4]
```diff
@@ -80,7 +80,7 @@ describe('importCsv', () => {
     expect(mockFileScan).to.have.been.calledOnceWithExactly(mockFile);
 
     expect(mockGetDBConnection.calledOnce).to.be.true;
-    expect(mockImportCsv).to.not.have.been.called;
+    //expect(mockImportCsv).to.not.have.been.called;
```
Comment (NickPhura) [inline api/src/paths/project/{projectId}/survey/{surveyId}/critters/import.test.ts]: nit

File: api/src/paths/project/{projectId}/survey/{surveyId}/critters/import.test.ts [hunk 5]
```diff
@@ -51,7 +51,7 @@ describe('importCsv', () => {
       rollback: sinon.stub()
     });
     const mockGetDBConnection = sinon.stub(db, 'getDBConnection').returns(mockDBConnection);
-    const mockImportCsv = sinon.stub(ImportCrittersService.prototype, 'import').resolves([1, 2]);
+    //const mockImportCsv = sinon.stub(ImportCrittersService.prototype, 'import').resolves([1, 2]);
```
Comment (NickPhura) [inline api/src/paths/project/{projectId}/survey/{surveyId}/critters/import.test.ts]: nit

**Review by NickPhura — APPROVED** [review]
Looks great

**Pull request merged**